src/man/olc.1

.TH olc 1 "29.4.2017" "version 0.1"
.SH NAME
olc - OutLander program for application of methodology of land evaluation for glider outlanding
.SH SYNOPSIS
.SY olc
.OP \-hslod
.OP \--l8
.OP \--s2
.BI \--isr " file"
.OP \--its file
.OP \--igt file
.OP \--ica name
.OP \--irr file
.OP \--demdir dir
.OP \--olr file
.OP \--occ file
.OP \--ocs file
.OP \--oca file
.OP \--out file
.OP \--oa file
.YS

.SH DESCRIPTION
This program applies methodology of land evaluation for glider outlanding. This methodology was tested using atmospherically corrected Sentinel-2 and Landsat 8 imagery. Program performs land cover classification using user defined training samples, classification quality assessment using user defined control sites and glider outlanding criteria applications using user defined reclassification file, DEM source and more.

.SH OPTIONS
.TP
.BI "-h, --help"
Prints help message and exits.

.TP
.BI "-s, --svm"
Use SVM classifier. By default, SVM with polynomial kernel and no optimization is used. LibSVM is used to perform actual training and classification.

.TP
.BI "-l, --linsvm"
Use SVM classifier without kernel (linear SVM). To use this option
.BI --svm
must be present.

.TP
.BI "-o, --opt"
Use SVM parameter optimization. Using this option results in heavy increase in computation time.

.TP
.BI "-d, --dt"
Use Decision tree classifier. Uses OpenCV library.

.TP
.BI "--l8"
Defines input satellite image source as Landsat 8. Needed for albedo calculation, otherwise useless. Input satellite raster file should contains atmospherically corrected bands 1-7 in 30 m resolution.

.TP
.BI "--s2"
Defines input satellite image source as Sentinel-2. Needed for albedo calculation, otherwise useless. Input satellite raster file should contains atmospherically corrected bands 2-7, 8B (referenced as 8A in metadata), 11 and 12 in 20 m resolution or 1-7, 8B (referenced as 8A in metadata), 9, 11 and 12 in 60 m resolution.

.TP
.BI "--isr" " file"
Specify input satellite image raster. This should by GeoTIFF including all sensor bands. If no albedo calculation is performed, image source or radiometrical pre-processing plays no role in program execution. In case of albedo calculation, satellite source must be defined, see
.BI "--l8"
and
.BI "--s2"
options.

.TP
.BI "--its" " file"
Specify input training samples used for classifier training. File should be ESRI ShapeFILE including polygon geometries representing training samples geometry and attribude describing class of sample. Coordinate system doesn't need to be the same as input satellite image.

.TP
.BI "--igt" " file"
Specify input control samples (ground truth) used for classification quality assessment. Form should be the same as for training samples.

.TP
.BI "--ica" " name"
Specify name of attribute that specifies class of training and control samples.

.TP
.BI "--irr" " file"
Text file containing reclassification rules for crop type and height criterion. File should contains two columns separated by space. First column defines source land cover category. Second column defines destination suitability category (0, 1 or 2).

.EE
Example:
.EE
1 0
.EE
2 1
.EE
3 2

.TP
.BI "--demdir" " dir"
Specify directory containing DEM tiles. At this time, only directory containing SRTM HGT files is suported.

.TP
.BI "--olr" " file"
Specify output land cover map GeoTIFF.

.TP
.BI "--occ" " file"
Specify output criterion of crop type and height map GeoTIFF.

.TP
.BI "--ocs" " file"
Specify output criterion of slope map GeoTIFF.

.TP
.BI "--oca" " file"
Specify output criterion of area map GeoTIFF.

.TP
.BI "--out" " file"
Specify output suitability map GeoTIFF.

.TP
.BI "--oa" " file"
Specify output albedo map GeoTIFF.

.SH AUTHOR
Peter Piesecký (pieseckypeter@posteo.net)


src/app/olc.cxx
#include "../oll/olcl.hxx"
#include <boost/program_options.hpp>
#include <ctime>
#include <iostream>
#include <string>

using std::string;
using std::cout;
using std::cerr;
using std::endl;

int main(int argc, char* argv[])
{
    // CLI creation
    namespace po = boost::program_options;

    // variables to hold CLI options
    string sourceImage, trainingSamples, groundTruth, classAtribure, demDir, reclasRulesFile, outRecl, outPodPlod, outPodSklon,
        outPodRozloh, outAll, outAlbedo;
    bool svm = false;
    bool svmlin = false;
    bool dt = false;
    int numClassifiers = 0;
    bool classify = false;
    bool reclassify = false;
    bool optimize = false;

    // CLI options declaration
    po::options_description desc("Allowed options");
    desc.add_options()("help,h", "produce help message")("svm,s", "Use SVM classifier.")("linsvm,l", "Use linear SVM classifier.")(
        "opt,o", "Optimize classifier parmetters.")("dt,d", "Use DT classifier.")("l8", "Satellite is L8")("s2", "Satellite is S2")(
        "isr", po::value< string >(&sourceImage), "Input multiband satellite raster data.")(
        "its", po::value< string >(&trainingSamples), "Input training samples vector.")("igt", po::value< string >(&groundTruth),
                                                                                        "Input ground truth vector.")(
        "ica", po::value< string >(&classAtribure), "Class attribute name in input training samples and ground truth vectors.")(
        "irr", po::value< string >(&reclasRulesFile),
        "Text file containing reclassification rules (1)")("demdir", po::value< string >(&demDir), "Directory containing SRTM hgt files.")(
        "olr", po::value< string >(&outRecl), "Output landcover raster.")("occ", po::value< string >(&outPodPlod),
                                                                          "Output condition of crop raster.")(
        "ocs", po::value< string >(&outPodSklon), "Output condition of slope raster.")("oca", po::value< string >(&outPodRozloh),
                                                                                       "Output condition of area raster.")(
        "out", po::value< string >(&outAll), "Output raster of suitability.")("oa", po::value< string >(&outAlbedo),
                                                                              "Output albedo raster.");

    string usage = " [-hslg] [--l8] [--s2] --isr --its --igt --ica --demdir [--irr] [--olr] [--occ] [--ocs] [--oa]";

    // CLI options handling
    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);
    bool parametersOk = true;

    if (vm.count("help"))
    {
        cout << "Usage: " << argv[0] << usage << endl << endl << desc << endl;
        return 1;
    }

    if (!vm.count("isr"))
    {
        cout << "Option isr is missing!" << endl;
        parametersOk = false;
    }

    if (vm.count("its") || vm.count("igt") || vm.count("ica") || vm.count("olr") || vm.count("occ") || vm.count("ocs") || vm.count("out"))
    {
        classify = true;
        if (!vm.count("svm") && !vm.count("dt"))
        {
            svm = dt = true;
            numClassifiers = 3;
            if (vm.count("linsvm"))
            {
                svmlin = true;
            }
        }
        else
        {
            if (vm.count("linsvm"))
            {
                svmlin = true;
            }

            if (vm.count("svm"))
            {
                svm = true;
                numClassifiers++;
            }

            if (vm.count("dt"))
            {
                dt = true;
                numClassifiers++;
            }
        }

        if (vm.count("opt"))
        {
            optimize = true;
        }

        if (!vm.count("its"))
        {
            cout << "Option its is missing!" << endl;
            parametersOk = false;
        }

        if (!vm.count("igt"))
        {
            cout << "Option igt is missing!" << endl;
            parametersOk = false;
        }

        if (!vm.count("ica"))
        {
            cout << "Option ica is missing!" << endl;
            parametersOk = false;
        }

        if (!vm.count("demdir"))
        {
            cout << "Option demdir is missing!" << endl;
            parametersOk = false;
        }

        if (!vm.count("olr") && !vm.count("occ") && !vm.count("ocs") && !vm.count("out"))
        {
            cout << "No output specified!" << endl;
            parametersOk = false;
        }

        if (vm.count("occ") || vm.count("ocs") || vm.count("out"))
        {
            reclassify = true;
        }
    }

    if (vm.count("oa"))
    {
        if ((!vm.count("l8") && !vm.count("s2")) || (vm.count("l8") && vm.count("s2")))
        {
            cout << "You need to specify satellite for albedo calculation!" << endl;
            parametersOk = false;
        }
    }

    if (!parametersOk)
    {
        cout << endl << "Use option -h [--help] for help." << endl << "Usage: " << argv[0] << usage << endl;
        return -1;
    }

    time_t timer1;
    time(&timer1);
    string sufix = std::to_string(timer1);
    GDALAllRegister();

    // input image reading
    if (!oll::checkIfExists(sourceImage, oll::inputFilePath))
    {
        cerr << "Can't read input satellite image!" << endl;
        return -1;
    }
    oll::VectorImageType::Pointer inputImage = oll::VectorImageType::New();
    oll::loadRaster(inputImage, sourceImage);

    if (classify)
    {
        // DEM registration
        otb::DEMHandler::Pointer demHandler = otb::DEMHandler::Instance();
        if (!demHandler->IsValidDEMDirectory(demDir.c_str()))
        {
            cerr << "Input dem directory is not usable!\n";
            return -1;
        }
        demHandler->OpenDEMDirectory(demDir);

        // input training samples reading
        if (!oll::checkIfExists(trainingSamples, oll::inputFilePath))
        {
            cerr << "Can't read input training samples!" << endl;
            return -1;
        }
        oll::VectorDataType::Pointer trainingSites = oll::VectorDataType::New();
        oll::loadVector(trainingSites, trainingSamples);

        // input ground truth vector reading
        if (!oll::checkIfExists(groundTruth, oll::inputFilePath))
        {
            cerr << "Can't read input ground truth!" << endl;
            return -1;
        }
        oll::VectorDataType::Pointer groundTruthVector = oll::VectorDataType::New();
        oll::loadVector(groundTruthVector, groundTruth);

        oll::ReclassificationRulesType reclassificationRules;
        if (reclassify)
        {
            if (!oll::checkIfExists(reclasRulesFile, oll::inputFilePath))
            {
                cerr << "Can't read input reclassification rules file!" << endl;
                return -1;
            }

            // reading and validation of input reclassification rules
            reclassificationRules = oll::readReclassificationRules(reclasRulesFile);
        }

        // image training and classification
        string modelDT = "/tmp/modelDT" + sufix;
        string modelSVM = "/tmp/modelSVM.txt" + sufix;
        oll::LabelImageType::Pointer DTClassified = oll::LabelImageType::New();
        oll::LabelImageType::Pointer SVMClassified = oll::LabelImageType::New();
        oll::confMatData DTcm, SVMcm;
        std::vector< oll::ConfusionMatrixType > matrices;
        std::vector< oll::ConfusionMatrixCalculatorType::MapOfClassesType > maps;
        oll::LabelImageListType::Pointer classifiedImages = oll::LabelImageListType::New();
        oll::LabelImageType::Pointer fusedImage;
        if (dt)
        {
            cout << "DT training ";
            cout.flush();

            oll::train(inputImage, trainingSites, modelDT, classAtribure, oll::desicionTree, false, optimize);
            cout << " classification  ";
            cout.flush();
            oll::classify(inputImage, modelDT, DTClassified);
            DTcm = oll::vypocitajChybovuMaticu(DTClassified, groundTruthVector, classAtribure);
            matrices.push_back(DTcm.confMat);
            maps.push_back(DTcm.mapOfClasses);
            classifiedImages->PushBack(DTClassified);
            fusedImage = DTClassified;

            cout << " DONE" << endl;

            oll::printConfMat(DTcm, cout);
        }

        if (svm)
        {
            cout << "SVM training ";
            cout.flush();

            oll::train(inputImage, trainingSites, modelSVM, classAtribure, oll::libSVM, svmlin, optimize);
            cout << " classification  ";
            oll::classify(inputImage, modelSVM, SVMClassified);
            SVMcm = oll::vypocitajChybovuMaticu(SVMClassified, groundTruthVector, classAtribure);
            matrices.push_back(SVMcm.confMat);
            maps.push_back(SVMcm.mapOfClasses);
            classifiedImages->PushBack(SVMClassified);
            fusedImage = SVMClassified;

            cout << " DONE" << endl;

            oll::printConfMat(SVMcm, cout);
        }

        if (numClassifiers > 1)
        {
            cout << "DS fusion ";
            cout.flush();

            // fusion of classified images
            oll::LabelImageType::Pointer fusedImage = oll::LabelImageType::New();
            oll::dsf(classifiedImages, matrices, maps, 0, 255, fusedImage);

            cout << " DONE" << endl;
        };

        if (vm.count("olr"))
        {
            oll::ulozRaster(fusedImage, outRecl);
            cout << "Classified raster  SAVED" << endl;
        }

        if (reclassify)
        {
            cout << "Crop type and height reclassification ";
            cout.flush();

            // condition of crop applicaiton
            oll::LabelImageType::Pointer podPlod = oll::LabelImageType::New();
            oll::reclassifyRaster(fusedImage, podPlod, reclassificationRules);

            cout << " DONE ";
            cout.flush();

            if (vm.count("occ"))
            {
                oll::ulozRaster(podPlod, outPodPlod);
                cout << "SAVED";
            }

            cout << endl;

            // condition of slope application
            if (vm.count("ocs") || vm.count("out") || vm.count("oca") || vm.count("out"))
            {
                cout << "Condition of slope ";
                cout.flush();

                oll::DoubleImageType::Pointer alignedDem = oll::DoubleImageType::New();
                oll::alignDEM(inputImage, alignedDem);

                oll::DoubleImageType::Pointer slope = oll::DoubleImageType::New();
                oll::computeSlopeRaster(alignedDem, slope);

                oll::LabelImageType::Pointer podSklon = oll::LabelImageType::New();
                oll::podSklon(podPlod, slope, podSklon, 8.13);

                cout << " DONE ";
                cout.flush();

                if (vm.count("ocs"))
                {
                    oll::ulozRaster(podSklon, outPodSklon);
                    cout << "SAVED";
                }

                cout << endl;

                // condition of area application
                cout << "Condition of area size ";
                cout.flush();

                oll::LabelImageType::Pointer podRozloh = oll::LabelImageType::New();
                oll::podRozloh(podSklon, podRozloh, 12000);
                cout << " DONE ";

                cout.flush();
                if (vm.count("oca"))
                {
                    oll::ulozRaster(podRozloh, outPodRozloh);
                    cout << "SAVED";
                }
                cout << endl;

                if (vm.count("out"))
                {
                    cout << "Suitability map reclassification ";
                    cout.flush();

                    oll::LabelImageType::Pointer suitable = oll::LabelImageType::New();
                    oll::mapOfSuitabilityCreation(fusedImage, podRozloh, suitable);
                    cout << " DONE ";
                    cout.flush();

                    oll::ulozRaster(suitable, outAll);
                    cout << " SAVED" << endl;
                }
            }
        }
    }

    if (vm.count("oa"))
    {
        cout << "Albedo calculation ";
        cout.flush();

        oll::satellites satellite;
        if (vm.count("l8"))
        {
            satellite = oll::Landsat8;
        }
        else if (vm.count("s2"))
        {
            satellite = oll::Sentinel2;
        }

        oll::DoubleImageType::Pointer albedo = oll::DoubleImageType::New();
        oll::albedo(inputImage, satellite, albedo);

        cout << " DONE ";
        cout.flush();

        oll::ulozRaster(albedo, outAlbedo);
        cout << " SAVED" << endl;
    }

    return 0;
}
